src/ldm.c

```
@@ -2729,6 +2729,121 @@ _dm_remove(const gchar * const name, uint32_t udev_cookie, GError ** const err)
     return r;
 }
 
+static int _find_partition(GString* target_params,
+    const gchar* dev_name, guint64 abs_part_start, guint64 part_size)
+{
+    char *dev_name_copy = strdup(dev_name); // basename might modify its arg
+    if (!dev_name_copy) {
+        return -1;
+    }
+    char *base_name = basename(dev_name_copy);
+    if (!base_name || strlen(base_name) == 0) {
+        free(dev_name_copy);
+        return -1;
+    }
+
+    char sysfs_path[PATH_MAX];
+    snprintf(sysfs_path, sizeof(sysfs_path), "/sys/block/%s", base_name);
+
+    DIR *dir = opendir(sysfs_path);
+    if (!dir) {
+        free(dev_name_copy);
+        return -1;
+    }
+
+    struct dirent *entry;
+    int found = 0;
+    while ((entry = readdir(dir)) != NULL && !found) {
```

You're not checking for errors from `readdir`.  That's quite complicated.  For a correct example see: https://github.com/libguestfs/libguestfs/blob/e49d65a5fd07e40260e800c0c262dad7e367b2f6/daemon/devsparts.c#L58-L89

---

Should I? The current implementation of _find_partition() fails if it's unable to match GPT/MBR partition to an LDM partition and readdir() errors actually have the same effect. I mean, both normal end of loop and readdir() error have the same outcome, so I don't explicitly differentiate them. Failure of _find_partition() doesn't mean the whole ldmtool call fails, as the _refine_partition() provides a fallback.

We can:
1. Leave it as is, i.e. silently fail and rely on fallback (that won't be silent if it fails).
2. Add error logging everywhere in _find_partition, including catching and logging and readdir() errors.

Thoughts?

---

Fixed.